### ISSUE TITLE
Access the current SqlConnection and SqlTransport from within an outgoing behavor

### ISSUE THREAD

**Issue by SimonCropp** [Particular/NServiceBus.SqlServer#455]
from within a behavor how do you access the SqlConnection and SqlTransport. specifically i want to re use them to write some extra data in the same context

**Comment by SzymonPobiega** [comment]
@SimonCropp behaviors have access to the context bag which contains the `TransportTransaction`. The `TransportTransaction` is another bag that contains `SqlConnection` and (in native tx mode) `SqlTransaction`.

**Comment by SimonCropp** [comment]
Can u add that to the doco. Also will that work from sending outside the pipeline. Eg in a web api controller

**Comment by SzymonPobiega** [comment]
For sending from outside of the endpoint you can use this API https://docs.particular.net/transports/sql/transactions#user-provided-transactions to set the connection/transaction. I am not sure if we want to document this approach using `TransportTransaction`. What's the usage scenario? I am concerned a bit about deprecating it in the future.

**Comment by SimonCropp** [comment]
@SzymonPobiega i want to dedup in the web tier.

**Comment by tmasternak** [comment]
@SimonCropp could you share a bit more details? Does the dedup mean:
 * atomicity of business operation and message send/publish e.g. in a controller,
 * deduplication of a message that can be physcially sent more than once,
 * something else that.

**Comment by SimonCropp** [comment]
So a web request comes in that will result in a message being sent. Message successfully sends, but connection cuts out and client never gets response. Most JS libs support a form of immediate retries. So post will come in again, and results in a duplicate message.

So we make the client specify a correlation id (a guid). Then we store that in a sql table when the message is sent from the web layer. If we see the same id again we dont send the message. 

The table is periodically cleaned of old ids

i pushed a prototype here https://github.com/NServiceBusExtensions/NServiceBus.Native/tree/master/SqlServer.Deduplication
there is a runnable sample and tests. but my current approach is flawed since my dedup runs prior to the message send, so not in the same db transaction.

basically i want this https://docs.microsoft.com/en-us/azure/service-bus-messaging/duplicate-detection

initially my use case is for "sending from outside the pipeline", eg web api, but i was planning on extending it to inside the pipeline as well

**Comment by SimonCropp** [comment]
thoughts?

**Comment by SzymonPobiega** [comment]
*Still thinking about this*

the use case seems legit. I assume the endpoints inside the system operate in exactly once mode and you need to dedup on the borders. You can't use outbox to do so because outbox *by design* introduces duplicates.

Adding public API to the `TransportTransaction` when using SQL Server transport that allows setting the connection/transaction makes sense to me, especially that we have similar API for getting them. 

What I wonder is if documenting that is the best way to go for now as it would be hard do obsolete. @tmasternak I'd rather add the new public API in a minor. What do you think?

**Comment by tmasternak** [comment]
@SimonCropp isn't the current [api](https://docs.particular.net/transports/sql/transactions#user-provided-transactions) enough to support deduplication for the web site? It allows for transaction spanning transport and your deduplication table.

When it comes to in-the-pipeline scenario it should be possible to use transport with TransactionScope and open a new connection in the behavior which (assuming same storage location) would not escalate to DTC.

Am I missing something here?

**Comment by SimonCropp** [comment]
>  It allows for transaction spanning transport and your deduplication table.

how would I do that?

**Comment by SimonCropp** [comment]
@tmasternak https://github.com/Particular/NServiceBus.SqlServer/issues/455#issuecomment-413046509

**Comment by tmasternak** [comment]
Inside the pipeline when using SqlTransport with TrascationScope transaction mode new connection openned in [SendBehavior](https://github.com/NServiceBusExtensions/NServiceBus.Native/blob/e33cf56de9aa8161977ce11b69141af1dbd20811/SqlServer.Deduplication/SendBehavior.cs#L24) will enroll in the wrapping transaction. Assuming that the connection string used inside `connectionBuilder` is the same as for the transport that will not escalate to DTC.

In the scenario when sending outside of the pipeline you could extend the [snippet](https://docs.particular.net/transports/sql/transactions#user-provided-transactions) by reusing native trasaction to check/update deduplication table.

**Comment by SimonCropp** [comment]
wont opening a new sql connection inside a send behaviour cause an escalation to dtc? https://petermeinl.wordpress.com/2011/03/13/avoiding-unwanted-escalation-to-distributed-transactions/

**Comment by andreasohlund** [comment]
>wont opening a new sql connection inside a send behaviour cause an escalation to dtc?

Yes it does, had a user hit this last week, raised this to track a potential solution 

https://github.com/Particular/NServiceBus.SqlServer/issues/457

The workaround is to skip using a scope,  is that viable for you @SimonCropp ?

**Comment by SimonCropp** [comment]
i am in the context of extending NSB. so i cant push that workaround on all people that use my component

**Comment by tmasternak** [comment]
@SimonCropp I think that there might be a workaround that would enable you to use the `TransactionScope` (all the credits go to @andreasohlund ;) ). The trick is to set `TransportTransaction` is such a way to force the transport to use native transaction and connection even when inside the scope. Here is asample:

``` C#
using (var scope = new TransactionScope(TransactionScopeAsyncFlowOption.Enabled))
{

    using (var connection = new SqlConnection(connectionString))
    {
        await connection.OpenAsync().ConfigureAwait(false);

        var guid = Guid.NewGuid().ToString();
        var commandTxt = $"<some-sql>";

        var command = new SqlCommand(commandTxt, connection);

        await command.ExecuteNonQueryAsync().ConfigureAwait(false);

        var options = new SendOptions();
        var transportTransaction = new TransportTransaction();
        transportTransaction.Set(connection);
        //transportTransaction.Set(scope);
        using (var beginTransaction = connection.BeginTransaction())
        {
            transportTransaction.Set(beginTransaction);

            options.GetExtensions().Set(transportTransaction);

            await endpointInstance.Send(orderAccepted, options).ConfigureAwait(false);

            beginTransaction.Commit();
        }
    }

    scope.Complete();
}
```

The important thing here is the native transaction and the way you set up the `TransportTransaction`.

**Comment by SimonCropp** [comment]
@tmasternak i am in the context of writing an extension to NSB. so can only plug into the pipeline. would that workaround work in that context? and if so wouldnt it break the documented "user-provided-transactions"

TBH i dont understand the pushback to adding an api/documented way for an extension to get access to the current connection and transaction?

**Comment by tmasternak** [comment]
@SimonCropp I re-read this issue and I think there indeed is some confusion from our end here. 

I understand that your aim is to provide `SqlServer.Deduplication` extension and the title of the issue is a solution that might enable it to work as intended. That being said I would really like to understand the broader context specifically what are the intended usages and how do you want the user code around sends to look.

The sample provided in the extension repo is a vanilla `IEndpointInstance .Send` which is probably not what would happen in the real-world scenario. The main issue being (as you already know) transaction management. So let me put in writing a short summary or the options to make sure we are all on the same page.

## Sends without any transactions
If the `Send` method on `IEndpointInstance` is called outside of `TransactionScope` and no native transaction has been passed via [UseCustomSqlConnectionAndTransaction]( https://docs.particular.net/transports/sql/transactions#user-provided-transactions) each send method will be executed in a separate transaction.

## Sends wrapped in `TransactionScope`
If the `Send` is wrapped in `TransactionScope` a new sql connection will be opened by the transport during message dispatch. This connection will enlist in the scope. If there is no other opened connection for the same scope there will be no DTC escalation happening.
In the context of your extension, you would have to change the way the behavior works and move [next() call](https://github.com/NServiceBusExtensions/NServiceBus.Native/blob/master/SqlServer.Deduplication/SendBehavior.cs#L43) outside of the using statement. Otherwise, the extension itself will cause two connections to be opened at the same time.

## Sends with user-provided transactions
If the `Send` is passed native transaction and connection using `SendOptions` those will be used by the dispatcher. There is indeed no documented way to get the custom connection and transaction options. The best thing I can propose for accessing those in the behavior is:
``` C#
context.Extensions.TryGet(out TransportTransaction transaction);

var sqlConnection = transaction.Get<SqlConnection>();
var sqlTransaction = transaction.Get<SqlTransaction>();
```
We might consider adding api for accessing those but the soonest it will happen is the next maintenance release which has no date confirmed yet.

In summary, I would like to point out that it might be hard (if not impossible) to make the featue work without DTC escalations for any code base as it boils down to connection management which depends on the end user. As a result the extension will need to provide some option for the users to provide a connection which will be reused by the extensions and the transport. Once more this is not documented but the [comment](https://github.com/Particular/NServiceBus.SqlServer/issues/455#issuecomment-421269664) above shows how to achieve this with the current state of affairs.

@SimonCropp please let me know if that sums up the state of this issue so far and if there is anything I have missed.

**Comment by SimonCropp** [comment]
> If the Send is wrapped in TransactionScope a new sql connection will be opened by the transport during message dispatch. This connection will enlist in the scope. If there is no other opened connection for the same scope there will be no DTC escalation happening. In the context of your extension, you would have to change the way the behavior works and move next() call outside of the using statement. Otherwise, the extension itself will cause two connections to be opened at the same time.

wouldnt exposing the transport connection and transaction work in this scenario?

> In summary, I would like to point out that it might be hard (if not impossible) to make the featue work without DTC escalations for any code base as it boils down to connection management which depends on the end user

perhaps i am missing something. but exposing the currently used connection and transaction to any api in the pipeline would seem to make this not too difficult? weather that conn and trans comes from user code or is initiated by the transport would seem irrelevant

**Comment by tmasternak** [comment]
Could you give a sample for the api that you would like to have? 

Is there anything this api would enable users to achieve what is not possible now? Or are you asking to make the content of `TransportTransaction` part of the public api and officially supported?
